File: kotlin/language/src/test/kotlin/sealedClasses.kt
import kotlin.test.Test
import kotlin.test.assertNotEquals
import kotlin.test.assertEquals

sealed class Action

abstract class Command : Action()
abstract class Query : Action()

fun handleAction(action: Action) : String {
    return when(action) {
        is Command -> "command"
        is Query -> "query"
    }
}

sealed class UserCommand : Command()

class CreateUserCommand(val nameToCreate : String) : UserCommand()

class DeleteUserCommand(val nameToDelete : String) : UserCommand()

fun handleUserCommands(action : UserCommand) : String {
    return when(action) {
        is CreateUserCommand -> "Create user ${action.nameToCreate}"
        is DeleteUserCommand -> "Delete user ${action.nameToDelete}"
    }
}

class SealedClassesTest {
    @Test fun `action handling`() {
        assertEquals( handleAction(CreateUserCommand("Lars")), "command" )
        assertEquals( handleAction(DeleteUserCommand("Lars")), "command" )
    }

    @Test fun `command handling`() {
        assertEquals( handleUserCommands(CreateUserCommand("Lars")), "Create user Lars" )
        assertEquals( handleUserCommands(DeleteUserCommand("Lars")), "Delete user Lars" )
    }
}

File: flutter/dart/sealed_classes.dart
sealed class Action { }

abstract class Command extends Action {} 

abstract class Query extends Action {}

String handleAction(Action action) {
    return switch(action) {
        Command() => "command",
        Query() => "query"
    };
}

sealed class UserCommand extends Command { }

class CreateUserCommand extends UserCommand {
  final String nameToCreate;
  CreateUserCommand(this.nameToCreate);
}

class DeleteUserCommand extends UserCommand {
  final String nameToDelete;
  DeleteUserCommand(this.nameToDelete);
}

String handleUserCommands(UserCommand action) {
    return switch(action) {
        CreateUserCommand() => "Create user ${action.nameToCreate}",
        DeleteUserCommand() => "Delete user ${action.nameToDelete}"
    };
}

main() {
  assertEquals( handleAction(CreateUserCommand("Lars")), "command" );
  assertEquals( handleAction(DeleteUserCommand("Lars")), "command" );

   assertEquals( handleUserCommands(CreateUserCommand("Lars")), "Create user Lars" );
  assertEquals( handleUserCommands(DeleteUserCommand("Lars")), "Delete user Lars" );
}

assertEquals(Object actual,Object expected) {
  if(actual == expected) {
    print("[$actual] matched [$expected]");
  } else {
    throw FormatException("[$actual] did not match [$expected]");
  }
}
